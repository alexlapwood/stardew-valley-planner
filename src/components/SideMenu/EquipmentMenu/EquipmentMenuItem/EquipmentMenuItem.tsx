import React from "react";

import cn from "classnames";

import { getSeason } from "../../../../helpers/date";
import BigText from "../../../BigText/BigText";
import Sprite from "../../../Sprite/Sprite";

// tslint:disable-next-line:no-var-requires
const { equipment, equipmentIds } = require("../../../../data/sdv.json") as {
  equipment: { [index: string]: IEquipment };
  equipmentIds: string[];
};

interface IProps {
  date: number;
  equipmentId: string;
  selectEquipment: (equipmentId: string, skinIndex: number) => void;
  selectedItem?: ISelectedItem;
}

interface IState {
  currentSkinIndex: number;
  open: boolean;
}

class EquipmentMenuItem extends React.Component<IProps, IState> {
  public state: IState = {
    currentSkinIndex: 0,
    open: false
  };

  public render() {
    return (
      <div>
        {this.renderPickItem(this.state.currentSkinIndex, true)}
        {this.state.open && this.renderPicklist()}
      </div>
    );
  }

  private close = () => {
    this.setState({ open: false });
  };

  private open = () => {
    this.setState({ open: true });
  };

  private toggle = () => {
    if (this.state.open) {
      this.close();
    } else {
      this.open();
    }
  };

  private renderPicklist = () => {
    const { equipmentId } = this.props;

    return equipment[equipmentId].skins.map((skinId, skinIndex) => (
      <div className="margin-left" key={skinId}>
        {this.renderPickItem(skinIndex)}
      </div>
    ));
  };

  private renderPickItem = (skinIndex: number, isTrigger?: boolean) => {
    const { equipmentId, date, selectEquipment, selectedItem } = this.props;

    let sprite: JSX.Element;

    let equipmentIndex = equipmentIds
      .slice(0, equipmentIds.indexOf(equipmentId))
      .reduce((acc, id) => {
        if (id === "fence" || id === "flooring") {
          return acc;
        }

        if (equipment[id].isSeasonal) {
          return acc + equipment[id].skins.length * 4;
        }

        return acc + equipment[id].skins.length;
      }, 0);

    if (equipment[equipmentId].isSeasonal) {
      equipmentIndex = equipmentIndex + skinIndex * 4 + getSeason(date);
    } else {
      equipmentIndex = equipmentIndex + skinIndex;
    }

    switch (equipmentId) {
      case "sprinkler":
        sprite = (
          <Sprite
            height={16}
            src="images/equipment-sheet.png"
            width={16}
            x={equipmentIndex * 16}
            y={16}
          />
        );
        break;
      case "fence":
        sprite = (
          <div
            style={{
              height: "16px",
              transform: "scale(0.5)",
              transformOrigin: "top left",
              whiteSpace: "nowrap",
              width: "16px"
            }}
          >
            <Sprite
              height={32}
              src="images/fences.png"
              width={16}
              x={skinIndex * 48}
              y={0}
            />
            <Sprite
              height={32}
              src="images/fences.png"
              width={16}
              x={skinIndex * 48 + 32}
              y={0}
            />
          </div>
        );
        break;
      case "flooring":
        sprite = (
          <div>
            <Sprite
              height={16}
              src="images/flooring.png"
              width={16}
              x={(skinIndex % 4) * 64}
              y={Math.floor(skinIndex / 4) * 64}
            />
          </div>
        );
        break;
      default:
        sprite = (
          <div
            style={{
              height: "16px",
              transform: "scale(0.5)",
              transformOrigin: "top center",
              whiteSpace: "nowrap",
              width: "16px"
            }}
          >
            <Sprite
              height={32}
              src="images/equipment-sheet.png"
              width={16}
              x={equipmentIndex * 16}
              y={0}
            />
          </div>
        );
        break;
    }
    return (
      <div
        className={cn("sdv-list-item", "flex-horizontal", {
          selected:
            selectedItem !== undefined &&
            selectedItem.type === "equipment" &&
            selectedItem.id === equipmentId &&
            selectedItem.skinIndex === skinIndex
        })}
        data-automationid={`equipment-${equipmentId}--${skinIndex}`}
        // tslint:disable-next-line:jsx-no-lambda
        onClick={() => {
          this.setState({ currentSkinIndex: skinIndex });
          selectEquipment(equipmentId, skinIndex);
        }}
      >
        <div className="sdv-list-item-icon">{sprite}</div>
        <div className="sdv-list-item-text flex margin-right">
          <BigText>{equipment[equipmentId].skins[skinIndex]}</BigText>
        </div>

        {isTrigger && equipment[equipmentId].skins.length > 1 && (
          <React.Fragment>
            <button
              className={cn("margin-right", {
                "sdv-button-down": !this.state.open,
                "sdv-button-up": this.state.open
              })}
              data-automationid="equipment-dropdown--trigger"
              // tslint:disable-next-line:jsx-no-lambda
              onClick={e => {
                e.stopPropagation();
                this.toggle();
              }}
            />
          </React.Fragment>
        )}
      </div>
    );
  };
}

export default EquipmentMenuItem;
