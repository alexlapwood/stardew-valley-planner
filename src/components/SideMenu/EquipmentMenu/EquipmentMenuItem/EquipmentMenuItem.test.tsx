import { mount } from "enzyme";
import React from "react";

import EquipmentMenuItem from "./EquipmentMenuItem";

// tslint:disable-next-line:no-var-requires
const { equipmentIds } = require("../../../../data/sdv.json") as {
  equipmentIds: string[];
};

describe("<EquipmentMenuItem />", () => {
  it("renders correctly", () => {
    const equipmentMenuItem = mount(
      <EquipmentMenuItem
        date={0}
        equipmentId={equipmentIds[0]}
        selectEquipment={jest.fn()}
      />
    );

    expect(equipmentMenuItem.find(".sdv-list-item")).toHaveLength(1);
    expect(equipmentMenuItem).toMatchSnapshot();
  });

  it("selects the item when clicked", () => {
    const selectEquipment = jest.fn();

    const equipmentMenuItem = mount<EquipmentMenuItem>(
      <EquipmentMenuItem
        date={0}
        equipmentId="scarecrow"
        selectEquipment={selectEquipment}
      />
    );

    equipmentMenuItem
      .find("[data-automationid='equipment-scarecrow--0']")
      .simulate("click");

    expect(selectEquipment).toHaveBeenCalledWith("scarecrow", 0);
  });

  describe("equipment with multiple skins", () => {
    it("shows extra skins when clicking the dropdown trigger", () => {
      const selectEquipment = jest.fn();
      const equipmentMenuItem = mount<EquipmentMenuItem>(
        <EquipmentMenuItem
          date={0}
          equipmentId="scarecrow"
          selectEquipment={selectEquipment}
        />
      );

      equipmentMenuItem
        .find("[data-automationid='equipment-dropdown--trigger']")
        .first()
        .simulate("click");

      expect(equipmentMenuItem.find(".sdv-list-item")).toHaveLength(10);
    });

    it("hides extra skins when clicking the dropdown trigger if the dropdown is already open", () => {
      const selectEquipment = jest.fn();
      const equipmentMenuItem = mount<EquipmentMenuItem>(
        <EquipmentMenuItem
          date={0}
          equipmentId="scarecrow"
          selectEquipment={selectEquipment}
        />
      );

      equipmentMenuItem.setState({ open: true });

      equipmentMenuItem
        .find("[data-automationid='equipment-dropdown--trigger']")
        .first()
        .simulate("click");

      expect(equipmentMenuItem.find(".sdv-list-item")).toHaveLength(1);
    });
  });

  it("updates the selected equipment when equipment is clicked", () => {
    const selectEquipment = jest.fn();
    const equipmentMenuItem = mount<EquipmentMenuItem>(
      <EquipmentMenuItem
        date={0}
        equipmentId="scarecrow"
        selectEquipment={selectEquipment}
      />
    );

    equipmentMenuItem.setState({ open: true });

    equipmentMenuItem
      .find("[data-automationid='equipment-scarecrow--1']")
      .simulate("click");

    expect(selectEquipment).toBeCalledWith("scarecrow", 1);
  });
});
