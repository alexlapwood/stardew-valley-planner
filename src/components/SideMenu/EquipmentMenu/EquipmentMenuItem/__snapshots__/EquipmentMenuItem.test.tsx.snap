// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EquipmentMenuItem /> renders correctly 1`] = `
<EquipmentMenuItem
  date={0}
  equipmentId="scarecrow"
  selectEquipment={[MockFunction]}
>
  <div>
    <div
      className="sdv-list-item flex-horizontal"
      data-automationid="equipment-scarecrow--0"
      onClick={[Function]}
    >
      <div
        className="sdv-list-item-icon"
      >
        <div
          style={
            Object {
              "height": "16px",
              "transform": "scale(0.5)",
              "transformOrigin": "top center",
              "whiteSpace": "nowrap",
              "width": "16px",
            }
          }
        >
          <Sprite
            height={32}
            src="images/equipment-sheet.png"
            width={16}
            x={0}
            y={0}
          >
            <div
              className="sprite"
              style={
                Object {
                  "backgroundImage": "url(images/equipment-sheet.png)",
                  "backgroundPosition": "-0px -0px",
                  "display": "inline-block",
                  "height": "32px",
                  "verticalAlign": "top",
                  "width": "16px",
                }
              }
            />
          </Sprite>
        </div>
      </div>
      <div
        className="sdv-list-item-text flex margin-right"
      >
        <BigText>
          <div
            className="BigText"
          >
            <div
              style={
                Object {
                  "transform": "scale(0)",
                  "transformOrigin": "left center",
                  "whiteSpace": "nowrap",
                }
              }
            >
              <Sprite
                height={16}
                key="0"
                src="images/sdv-font.png"
                width={8}
                x={408}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-408px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="1"
                src="images/sdv-font.png"
                width={8}
                x={536}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-536px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="2"
                src="images/sdv-font.png"
                width={8}
                x={520}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-520px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="3"
                src="images/sdv-font.png"
                width={8}
                x={656}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-656px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="4"
                src="images/sdv-font.png"
                width={8}
                x={552}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-552px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="5"
                src="images/sdv-font.png"
                width={8}
                x={536}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-536px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="6"
                src="images/sdv-font.png"
                width={8}
                x={656}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-656px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="7"
                src="images/sdv-font.png"
                width={8}
                x={632}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-632px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
              <Sprite
                height={16}
                key="8"
                src="images/sdv-font.png"
                width={8}
                x={696}
                y={0}
              >
                <div
                  className="sprite"
                  style={
                    Object {
                      "backgroundImage": "url(images/sdv-font.png)",
                      "backgroundPosition": "-696px -0px",
                      "display": "inline-block",
                      "height": "16px",
                      "verticalAlign": "top",
                      "width": "8px",
                    }
                  }
                />
              </Sprite>
            </div>
          </div>
        </BigText>
      </div>
      <button
        className="margin-right sdv-button-down"
        data-automationid="equipment-dropdown--trigger"
        onClick={[Function]}
      />
    </div>
  </div>
</EquipmentMenuItem>
`;
